### PR TITLE
fix: Allow home directory path in workflow project config

### PR DESCRIPTION
**Pull request by a-li**
Issue #, if available:

**Description of Changes**
Allow home directory path in workflow project config.

This is the current behavior before my change:

```
name: Demo
schemaVersion: 1
workflows:
  hello:
    type:
      language: wdl
      version: 1.0
    sourceURL: ~/work/agc/amazon-genomics-cli/examples/demo-wdl-project/workflows/hello/hello.wdl
```
However, this will not work since `~` is not correctly expanded when running workflow run, resulting in a file not found error:
```
2021-11-01T16:25:22-07:00 𝒊  Running workflow. Workflow name: 'hello', Arguments: '', Context: 'myContext'
2021-11-01T16:25:23-07:00 ✘   error="unable to run workflow: file '~/work/agc/amazon-genomics-cli/examples/demo-wdl-project/workflows/hello/hello.wdl' doesn't exist"
Error: an error occurred invoking 'workflow run'
with variables: {WorkflowName:hello Arguments: ContextName:myContext}
caused by: unable to run workflow: file '~/work/agc/amazon-genomics-cli/examples/demo-wdl-project/workflows/hello/hello.wdl' doesn't exist
```

After my change, it is able to successfully run the workflow
```
2021-11-01T16:29:14-07:00 𝒊  Running workflow. Workflow name: 'hello', Arguments: '', Context: 'myContext'
7b2489b4-9816-48bb-b7f8-84bc2023f33b
```

[//]: #  (A description of the change that you made and the new user experience that it creates)

**Description of how you validated changes**
* Build
* Ran unit tests
* Manually modified the agc-project.yaml file to contain workflow with sourceURl starting with `~` and ran `workflow run ...` with the change.

[//]: #  (A description of you validated your changes and any relevant logs that show your change works)

**Checklist**

- [x] If this change would make any existing documentation invalid, I have included those updates within this PR
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] I have linted my code before raising the PR
- [x] Title of this Pull Request follows Conventional Commits standard: https://www.conventionalcommits.org/en/v1.0.0/

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go [hunk 1]
```diff
@@ -25,6 +27,14 @@ func CompressToTmp(srcPath string) (string, error) {
 }
 
 func writeToZipRecursive(writer *zip.Writer, rootPath string) error {
+	// Expand home directory path
+	usr, _ := user.Current()
```
Comment (elliot-smith) [inline packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go:31]: We probably don't want to ignore this error as it will cause issues if the customer is using home path and the Current() function fails but we also don't want to throw an error if the prefix doesn't exist in the path definition.

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go [hunk 2]
```diff
@@ -25,6 +27,14 @@ func CompressToTmp(srcPath string) (string, error) {
...
+	dir := usr.HomeDir
+	if rootPath == "~" {
```
Comment (elliot-smith) [inline packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go:33]: Can you please add some tests for this? There are some existing examples you can extend although you will need to create a global function for `user.Current` and `usr.HomeDir` to test this (i.e. Like the mkDirTemp example here: https://github.com/aws/amazon-genomics-cli/blob/main/packages/cli/internal/pkg/aws/cdk/deploy_app.go#L9 ).

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go [hunk 3]
```diff
@@ -24,16 +27,33 @@ func CompressToTmp(srcPath string) (string, error) {
 	return packFile.Name(), nil
 }
 
+func expandHomeDir(rootPath string) (string, error) {
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go:30]: This method should be moved to another package (can be useful outside zip package).
Options:
1) move it to `config` package
2) crate a new package `osutils` and move both DetermineHomeDir and ExpandHomeDir to this new package

**Review by IllyaYalovyy — CHANGES_REQUESTED** [review]

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go (hunk 3 above)
Comment (a-li) [reply packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go]: I created an `osutils` package and moved the relevant functions there.

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go [hunk 4]
```diff
@@ -24,16 +27,33 @@ func CompressToTmp(srcPath string) (string, error) {
...
+	homeDir, err := config.DetermineHomeDir()
+	if rootPath == "~" {
+		rootPath = homeDir
+	} else if strings.HasPrefix(rootPath, "~/") {
```
[the remark this answers is not in the document]
Comment (a-li) [reply packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go]: Addressed.

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go [hunk 5]
```diff
@@ -24,16 +27,33 @@ func CompressToTmp(srcPath string) (string, error) {
 	return packFile.Name(), nil
 }
 
+func expandHomeDir(rootPath string) (string, error) {
+	homeDir, err := config.DetermineHomeDir()
```
[the remark this answers is not in the document]
Comment (a-li) [reply packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go]: Addressed.

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go (hunk 2 above)
Comment (a-li) [reply packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go]: Added unit tests.

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go (hunk 1 above)
Comment (a-li) [reply packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go]: Changed it to use the existing `DetermineHomeDir` which provides error handling.

File: packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go [hunk 6]
```diff
@@ -24,16 +27,33 @@ func CompressToTmp(srcPath string) (string, error) {
...
+	if rootPath == "~" {
+		rootPath = homeDir
+	} else if strings.HasPrefix(rootPath, "~/") {
+		rootPath = filepath.Join(homeDir, rootPath[2:])
+	} else {
+		err = nil
```
[the remark this answers is not in the document]
Comment (a-li) [reply packages/cli/internal/pkg/cli/zipfile/compress_to_tmp.go]: Addressed.

File: packages/cli/internal/pkg/osutils/utils_test.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	expectedExpandedPath := "home/user/FooBar"
+
+	mockOs.EXPECT().UserHomeDir().Return(expectedHomePath, nil)
+	actualExpandedPath, _ := ExpandHomeDir("~/FooBar")
+
+	assert.Equal(t, expectedExpandedPath, actualExpandedPath)
+	ctrl.Finish()
+}
+
+func TestExpandHomeDir_WithoutExpansion(t *testing.T) {
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/osutils/utils_test.go:53]: 1. I think for testing this function it would be great to employ parameterized tests (more info on that: https://gobyexample.com/testing) We have a bunch of good examples as well: (https://github.com/aws/amazon-genomics-cli/blob/119c6303ecfec6b6ac346e3fcdc22a364abba533/packages/cli/internal/pkg/cli/describe_output_test.go)
2. Please add more test cases  for edge cases (e.g. `~` in random places of the path, one can create an actual folder with this name. `\~\~\~/`,  `~\~\~`, `foo~/~bar`, `~foo/bar~`, etc)
3. Please add a _separate_ test for error handling.

**Review by IllyaYalovyy — CHANGES_REQUESTED** [review]

File: packages/cli/internal/pkg/osutils/utils_test.go (hunk 7 above)
Comment (IllyaYalovyy) [reply packages/cli/internal/pkg/osutils/utils_test.go:53]: One more test case to add `~`.

Comment (IllyaYalovyy) [reply packages/cli/internal/pkg/osutils/utils_test.go:53]: Just an empty string as an input is another good test case

Comment (a-li) [reply packages/cli/internal/pkg/osutils/utils_test.go]: Added additional unit tests to cover the edge cases. Also utilized parameterized tests for a portion of the unit tests as suggested.

File: packages/cli/internal/pkg/osutils/utils.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		return "", actionableerror.New(err, "Please check that your home or user profile directory is defined within your environment variables")
+	}
+	return dir, nil
+}
+
+// ExpandHomeDir returns the expanded home directory path for the current user
+func ExpandHomeDir(rootPath string) (string, error) {
+	if rootPath == "~" {
+		homeDir, _ := DetermineHomeDir()
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/osutils/utils.go:26]: Why are we ignoring error in this case?

File: packages/cli/internal/pkg/osutils/utils_test.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	tests := map[string]struct {
+		pathTobeExpanded string
+		expectedPath     string
+	}{
+		"Tilde at the beginning": {
+			pathTobeExpanded: "~~~",
+			expectedPath:     "~~~",
+		},
+		"Tilde at the beginning with escape and backslash": {
+			pathTobeExpanded: "\\~\\~\\~/",
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/osutils/utils_test.go:107]: Sorry. what I actually meant was `/~/~/~/` and `~/~/~`, etc. I'm not sure what happen to formatting of my comment. Probably mixed up markdown. Back-slash can be part of the path, but I don't think it will affect current logic in any way.

File: packages/cli/internal/pkg/osutils/utils_test.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+		},
+		"Empty string": {
+			pathTobeExpanded: "",
+			expectedPath:     "",
+		},
+		"Relative": {
+			pathTobeExpanded: "FooBar",
+			expectedPath:     "FooBar",
+		},
+		"Absolute": {
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/osutils/utils_test.go:130]: I don't see any test cases that actually expand a path. for all test cases pathTobeExpanded == expectedPath. Am I missing anything?

**Review by IllyaYalovyy — CHANGES_REQUESTED** [review]

File: packages/cli/internal/pkg/osutils/utils.go (hunk 8 above)
Comment (a-li) [reply packages/cli/internal/pkg/osutils/utils.go:26]: Ooops must be slipped after rebasing the branch. It's changed now.

File: packages/cli/internal/pkg/osutils/utils_test.go (hunk 10 above)
Comment (a-li) [reply packages/cli/internal/pkg/osutils/utils_test.go:130]: Yeah these ones are not supposed to be expanded as it's not `~` nor `~/...`. I'll change the variable name to input so it's less confusing. The ones to be expanded are tested in unit tests between line 38 and 81.

File: packages/cli/internal/pkg/osutils/utils_test.go (hunk 9 above)
Comment (a-li) [reply packages/cli/internal/pkg/osutils/utils_test.go:107]: Cool, added these ones to the test as well.

File: packages/cli/internal/pkg/osutils/utils_test.go (hunk 10 above)
Comment (a-li) [reply packages/cli/internal/pkg/osutils/utils_test.go:130]: Synced offline. Will move all positive cases into a parameterized test.

**Review by AndreyDovydenko — APPROVED** [review]

**Review by elliot-smith — APPROVED** [review]

**Review by IllyaYalovyy — APPROVED** [review]
